### PR TITLE
Stream CDM CloudWatch logs to S3 with SNS/SQS notifications

### PR DESCRIPTION
**Pull request by dav3r**
#  #

## 🗣 Description ##

This PR sets up a process to stream the data in the [CDM CloudWatch log groups](https://github.com/cisagov/cool-sharedservices-cdm/blob/45710e672e52966348d7642250efdda3e319360d/variables.tf#L71-L75) to an S3 bucket so that it can be ingested by the CDM system.  It includes an SNS topic that receives notifications when new objects are added to the S3 bucket, as well as an SQS queue that is subscribed to the SNS topic.  The SNS topic and SQS queues are [required by the tool](https://docs.splunk.com/Documentation/AddOns/released/AWS/SQS-basedS3) that CDM plans to use for ingesting this data.

## 💭 Motivation and context ##

The tool used by CDM is not able to assume a role to ingest the CloudWatch logs directly from CloudWatch.  However it is able to assume a role to ingest the data if it is in an S3 bucket with an SNS topic and SQS queue as defined in this PR.  The data flow looks like this:

```mermaid
flowchart LR
    A(CloudWatch Log Groups with subscription filters) -->|Amazon Data Firehose| B(S3 Bucket)
    B -->|S3 Notifications| C(SNS Topic)
    C -->|Subscription| D(SQS Queue)
    C -->|Undeliverable Messages| E[SQS Dead Letter Queue]
```

Note that this PR does **not** include the permissions changes required for the CDM user to access the S3 bucket, SNS topic, and SQS queue - that will be handled in a separate PR (currently in #45, but subject to change).

> [!NOTE]
>  ~The `blocked` tag has been added to this PR since it is now dependent on https://github.com/cisagov/cool-cdm-cloudtrail-tf-module/pull/31.  This note and the `blocked` tag will be removed when https://github.com/cisagov/cool-cdm-cloudtrail-tf-module/pull/31 is merged.~

## 🧪 Testing ##

I successfully applied these changes with no errors in the Staging Shared Services account, then I confirmed all of the following:
 * Requests are flowing through the Firehose stream and being delivered to the S3 bucket (visible on the "Monitoring" tab of the Firehose web console)
 * Public access to the bucket is disabled (visible on the "Permissions" tab of the S3 web console)
 * The S3 bucket is encrypted with server-side encryption with Amazon S3 managed keys (SSE-S3) (visible on the "Properties" tab of the S3 web console)
 * Objects are in the S3 bucket and being sorted in folders by date and hour (visible in the "Objects" tab of the S3 web console)
 * The S3 objects contain correct message data from the CDM log groups/streams (verified by downloading a few and looking at their content)
 * The S3 bucket is correctly configured to notify the SNS topic for "All object create events" (visible on the "Properties" tab of the S3 web console)
 * The SNS topic is deployed with the correct access policy allowing only the S3 bucket to publish to the topic (visible in the "Access policy" tab of the SNS web console)
 * The SQS queue is subscribed to the SNS topic and the subscription is confirmed (visible in the "Subscriptions" tab of the SNS web console as well as in the "SNS subscriptions" tab of the SQS web console)
 * Messages are continuously arriving in the SQS queue, awaiting pickup by the CDM tool (visible in the "Monitoring" tab of the SQS web console)
 * The SQS queue is using the correct SQS dead-letter queue (visible in the "Dead-letter queue" tab of the SQS web console)
 * Both SQS queues (main and dead-letter) are encrypted with server-side encryption, managed by SSE SQS (visible in the "Encryption" tab of the SQS web console)
 * Both SQS queues (main and dead-letter) are deployed with the correct access policy allowing only the SNS topic to send messages to the queue (visible in the "Access policy" tab of the SQS web console)

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##
- [x] Revert testing branch commit https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/a079104e492868faa1a4b120d64a62f65fc5890e

## ✅ Post-merge checklist ##

- [ ] Apply these changes in Production
- [ ] Work with CDM engineers to ensure that the CDM role has the correct permissions to successfully fetch the CloudWatch data from the S3 bucket (will be handled as part of #45)

### CODE REVIEW

**Review by jsf9k — APPROVED** [review]
Nice work!  The diagram is a nice touch.

**Review by jsf9k — APPROVED** [review]
It's so nice I approved it twice!

File: README.md
Comment (mcdonnnj) [inline README.md:1]: Please update the repository description in the README to mention this pretty significant additional functionality.

File: cloudwatch_logs_for_cdm.tf [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+# Create CloudWatch subscription filters to send events from CDM CloudWatch
+# log groups to the Firehose delivery stream
+resource "aws_cloudwatch_log_subscription_filter" "cdm" {
+  provider = aws.sharedservicesprovisionaccount
+  for_each = toset(var.cloudwatch_policy_log_groups)
+
+  destination_arn = aws_kinesis_firehose_delivery_stream.cloudwatch_logs.arn
+  filter_pattern  = ""
```
Comment (mcdonnnj) [inline cloudwatch_logs_for_cdm.tf:50]: Might be worth mentioning in a comment that this matches everything.

File: cloudwatch_to_firehose_role.tf [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  statement {
+    actions = [
+      "firehose:PutRecord",
+      "firehose:PutRecordBatch",
+    ]
+
+    effect = "Allow"
+
+    resources = [
+      "arn:aws:firehose:${var.aws_region}:${local.sharedservices_account_id}:deliverystream/${var.firehose_delivery_stream_name}",
```
Comment (mcdonnnj) [inline cloudwatch_to_firehose_role.tf:51]: Are we unable to use `aws_kinesis_firehose_delivery_stream.cloudwatch_logs.arn`? I prefer not to handcraft ARNs unless specifically necessary.

File: cloudwatch_sns.tf [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    resources = [
+      aws_sns_topic.cloudwatch_logs.arn,
+    ]
+  }
+}
+
+# Attach the policy (that allows the CDM S3 bucket to send notifications to the
+# SNS topic) to the SNS topic.
```
Comment (mcdonnnj) [inline cloudwatch_sns.tf]: I get why they're here but can this be reworded to remove the parentheses?

File: variables.tf [hunk 4]
```diff
@@ -56,6 +61,18 @@ variable "aws_region" {
   default     = "us-east-1"
 }
 
+variable "cloudwatch_logs_sqs_queue_name" {
```
Comment (mcdonnnj) [inline variables.tf:64]: It might be worth mentioning that this is also used as a base for the dead letter queue's name.

**Review by mcdonnnj — APPROVED** [review]
This is looking pretty solid. Strong work 💪💪💪 I have some minor requests/suggestions for your consideration.

File: README.md
Comment (dav3r) [reply README.md:1]: Please have a look at https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/354b667092685a3e88178bca9f8ac281901c9eff.

File: cloudwatch_logs_for_cdm.tf (hunk 1 above)
Comment (dav3r) [reply cloudwatch_logs_for_cdm.tf:50]: Good thinking - I expanded the comment in https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/762a214980a3dc92e3ac993a683ada9d3a91c585.

File: cloudwatch_to_firehose_role.tf (hunk 2 above)
Comment (dav3r) [reply cloudwatch_to_firehose_role.tf:51]: We cannot due to a circular dependency.  I added an explanatory comment in https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/c3748705bf3573ee9af6b9ec58ed4c4c8f8a13ed.

File: cloudwatch_sns.tf (hunk 3 above)
Comment (dav3r) [reply cloudwatch_sns.tf]: Try https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/c4edbc67d4dad15f8b322bba74f0ca6c97150924 on for size and let me know what you think.

File: variables.tf (hunk 4 above)
Comment (dav3r) [reply variables.tf:64]: Good idea.  I did that in https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/8df490c3610a1e649c0dcd48d72c0422445064f2.

**Review by jsf9k — APPROVED** [review]

File: cloudwatch_sns.tf (hunk 3 above)
Comment (mcdonnnj) [reply cloudwatch_sns.tf]: Seems like a good fit!

**Pull request merged**